File: generator.py
#!/usr/bin/env python3
"""cfg2yacc generator module.

This script transforms a ``.def`` grammar specification into the collection of
generated artefacts the build expects:

* ``lexer.l`` – Flex rules derived from ``%%LEX``
* ``parser.y`` – Bison grammar synthesised from ``%%YACC``
* ``*_tokens.txt`` – token samples harvested from the analyzer's input file

It is invoked from the ``Makefile`` as part of ``make all`` / ``make run`` and
acts as the bridge between the high-level declarative grammar and the concrete
Flex/Bison sources compiled into ``custom_compiler``.  Every public function
carries a docstring so the file doubles as living documentation for the
generation pipeline.
"""

import sys
import re
from pathlib import Path
from typing import List, Tuple, Optional


class LexRule:
    def __init__(self, token_name: str, regex: str):
        self.token_name = token_name
        self.regex = regex

    # ``LexRule`` mirrors one line from the ``%%LEX`` block. ``token_name`` is
    # the symbolic token the parser sees while ``regex`` is the original
    # Flex-compatible pattern copied verbatim into ``lexer.l``.

class GrammarRule:
    def __init__(self, lhs: str, rhs: str, action: str):
        self.lhs = lhs
        self.rhs = rhs
        self.action = action

    # ``GrammarRule`` represents a single production taken from ``%%YACC``. The
    # left-hand side non-terminal is stored in ``lhs``, the textual production
    # in ``rhs`` and any associated semantic action (from ``{ ... }``) in
    # ``action``.

def parse_def_file(filename: str) -> Tuple[List[LexRule], List[GrammarRule]]:
    """Parse a ``.def`` analyzer file into lexer and grammar structures.

    Expects exactly one ``%%LEX`` followed by one ``%%YACC`` section.  The
    approach is intentionally pragmatic: it scans lines, extracts token /
    regex pairs, and records grammar productions including simple semantic
    actions.  Extremely elaborate multi-line actions may require manual touch
    ups, but typical samples are supported.
    """

    with open(filename, 'r') as f:
        content = f.read()

    # Basic validation that the required sections exist
    if '%%LEX' not in content or '%%YACC' not in content:
        print(f'Error: {filename} must contain both %%LEX and %%YACC sections')
        sys.exit(1)

    # Split the file into lex and yacc sections. This assumes a single
    # `%%LEX` followed by a single `%%YACC` in that order.
    parts = content.split('%%LEX')
    after_lex = parts[1].split('%%YACC')
    lex_section = after_lex[0]
    yacc_section = after_lex[1]
    
    lex_rules = []
    for line in lex_section.strip().split('\n'):
        line = line.strip()
        if not line or line.startswith('#'):
            continue
        parts = line.split(None, 1)
        if len(parts) == 2:
            token_name, regex = parts
            lex_rules.append(LexRule(token_name, regex))
    
    grammar_rules = []
    lines = yacc_section.strip().split('\n')
    i = 0
    while i < len(lines):
        line = lines[i].strip()
        
        if not line or line.startswith('#'):
            i += 1
            continue
        
        if '->' in line:
            parts = line.split('->', 1)
            lhs = parts[0].strip()
            rest = parts[1].strip()
            
            if '{' in rest:
                brace_content = rest[rest.index('{')+1:]
                rhs = rest[:rest.index('{')].strip()
                action = brace_content[:brace_content.index('}')].strip()
            else:
                rhs = rest
                action = ''
                if i + 1 < len(lines):
                    next_line = lines[i + 1].strip()
                    if next_line.startswith('{'):
                        action = next_line[next_line.index('{')+1:next_line.index('}')].strip()
                        i += 1
            
            grammar_rules.append(GrammarRule(lhs, rhs, action))
        
        elif line.startswith('|'):
            if grammar_rules:
                last_lhs = grammar_rules[-1].lhs
                rest = line[1:].strip()
                
                if '{' in rest:
                    brace_content = rest[rest.index('{')+1:]
                    rhs = rest[:rest.index('{')].strip()
                    action = brace_content[:brace_content.index('}')].strip()
                else:
                    rhs = rest
                    action = ''
                    if i + 1 < len(lines):
                        next_line = lines[i + 1].strip()
                        if next_line.startswith('{'):
                            action = next_line[next_line.index('{')+1:next_line.index('}')].strip()
                            i += 1
                
                grammar_rules.append(GrammarRule(last_lhs, rhs, action))
        
        i += 1
    
    return lex_rules, grammar_rules


def flex_regex_to_python(pattern: str) -> str:
    """Convert a Flex regex into a Python-compatible pattern string."""

    converted = pattern.strip()
    # Flex uses escaped forward slashes for literal '/', translate to Python style.
    converted = converted.replace('\/', '/')
    converted = converted.replace('\\"', '"')

    # Surround bare patterns with a non-capturing group to avoid precedence issues
    if converted and not converted.startswith('('):
        converted = f'(?:{converted})'

    return converted


def find_input_file(def_path: Path) -> Optional[Path]:
    """Infer the most appropriate sample input file for an analyzer."""

    # Common convention: `S1_analyzer.def` -> `S1_input.txt`
    base_name = def_path.stem
    candidates = []

    if base_name.endswith('_analyzer'):
        candidates.append(def_path.with_name(base_name.replace('_analyzer', '_input') + '.txt'))

    # Same-name .txt fallback (e.g., my_grammar.def -> my_grammar.txt)
    candidates.append(def_path.with_suffix('.txt'))

    # Search sibling files that contain 'input' in their stem for best effort
    for candidate in def_path.parent.glob('*'):
        if candidate.is_file() and candidate.suffix.lower() == '.txt' and 'input' in candidate.stem.lower():
            candidates.append(candidate)

    for candidate in candidates:
        if candidate.exists():
            return candidate

    return None

def generate_lexer(lex_rules: List[LexRule], output_file: str):
    """Emit a Flex ``lexer.l`` implementation from parsed ``LexRule`` entries."""

    with open(output_file, 'w') as f:
        # C prologue required by flex/bison integration
        f.write('%{\n')
        f.write('#include <stdio.h>\n')
        f.write('#include <stdlib.h>\n')
        f.write('#include <string.h>\n')
        f.write('#include "ast.h"\n')
        f.write('#include "y.tab.h"\n')
        f.write('%}\n\n')
        f.write('%%\n\n')

        # Emit each lexer rule. The ``.regex`` is written verbatim; the .def
        # author is responsible for providing Flex-compatible patterns.
        for rule in lex_rules:
            if rule.token_name == 'WHITESPACE':
                # Skip whitespace tokens entirely (no AST node created)
                f.write(rule.regex + '    { /* skip whitespace */ }\n')
            else:
                # For named tokens, create a leaf node and return the token
                f.write(rule.regex + '    { ')
                f.write('yylval.node = create_leaf_node("' + rule.token_name + '", yytext); ')
                f.write('return ' + rule.token_name + '; ')
                f.write('}\n')

        # Fallback rule for unexpected input characters
        f.write('\n.    { fprintf(stderr, "Unexpected character: %s\\n", yytext); }\n')
        f.write('%%\n\n')
        f.write('int yywrap() { return 1; }\n')

def generate_parser(lex_rules: List[LexRule], grammar_rules: List[GrammarRule], output_file: str):
    """Produce a Bison ``parser.y`` using the collected rule definitions."""

    with open(output_file, 'w') as f:
        # Bison C prologue: includes and forward declarations
        f.write('%{\n')
        f.write('#include <stdio.h>\n')
        f.write('#include <stdlib.h>\n')
        f.write('#include <string.h>\n')
        f.write('#include "ast.h"\n\n')
        f.write('extern int yylex();\n')
        f.write('extern int yyparse();\n')
        f.write('extern FILE *yyin;\n')
        f.write('void yyerror(const char *s);\n\n')
        f.write('Node *ast_root = NULL;\n')
        f.write('%}\n\n')
        f.write('%union {\n')
        f.write('    Node *node;\n')
        f.write('}\n\n')

        # Gather token names from lex rules (skip WHITESPACE)
        tokens = set()
        for rule in lex_rules:
            if rule.token_name != 'WHITESPACE':
                tokens.add(rule.token_name)

        # Emit %token declarations with the Node* semantic type
        for token in sorted(tokens):
            f.write('%token <node> ' + token + '\n')
        f.write('\n')

        # Declare %type for each nonterminal
        nonterminals = set()
        for rule in grammar_rules:
            nonterminals.add(rule.lhs)

        for nt in sorted(nonterminals):
            f.write('%type <node> ' + nt + '\n')
        f.write('\n')

        # Set the start symbol to the first declared LHS
        if grammar_rules:
            f.write('%start ' + grammar_rules[0].lhs + '\n\n')

        f.write('%%\n\n')

        # Emit grammar rules grouped by LHS, using '|' for alternatives and
        # preserving any user-provided actions.
        current_lhs = None
        is_first_rule = True
        for rule in grammar_rules:
            if rule.lhs != current_lhs:
                if current_lhs is not None:
                    f.write('    ;\n\n')
                f.write(rule.lhs + ':\n')
                current_lhs = rule.lhs
                first = True
            else:
                f.write('    | ')
                first = False

            if first:
                f.write('    ')

            f.write(rule.rhs if rule.rhs else '/* empty */')

            # Attach action code if present. For the very first grammar rule
            # record the root AST node in ``ast_root``.
            if rule.action:
                f.write('\n        { ' + rule.action)
                if is_first_rule:
                    f.write(' ast_root = $$;')
                f.write(' }')
            elif is_first_rule:
                f.write('\n        { ast_root = $$; }')

            f.write('\n')
            is_first_rule = False

        if current_lhs is not None:
            f.write('    ;\n\n')

        # Epilogue: error handler and a small main() that runs the parser
        f.write('%%\n\n')
        f.write('void yyerror(const char *s) {\n')
        f.write('    fprintf(stderr, "Parse error: %s\\n", s);\n')
        f.write('}\n\n')
        f.write('int main(int argc, char **argv) {\n')
        f.write('    if (argc > 1) {\n')
        f.write('        FILE *file = fopen(argv[1], "r");\n')
        f.write('        if (!file) {\n')
        f.write('            perror(argv[1]);\n')
        f.write('            return 1;\n')
        f.write('        }\n')
        f.write('        yyin = file;\n')
        f.write('    }\n\n')
        f.write('    int result = yyparse();\n\n')
        f.write('    if (result == 0 && ast_root != NULL) {\n')
        f.write('        printf("\\n=== Parse Tree ===\\n");\n')
        f.write('        print_ast(ast_root, 0);\n')
        f.write('        free_ast(ast_root);\n')
        f.write('    }\n\n')
        f.write('    return result;\n')
        f.write('}\n')

def generate_token_files(lex_rules: List[LexRule], def_file: str):
    """
    Generate token example files by scanning the analyzer's sample input.

     Steps performed:
     1. Locate the input file associated with the `.def` analyzer.
     2. Remove any previously generated `*_tokens.txt` files so that output is
         always derived from the latest input.
     3. For each lex rule (excluding punctuation/whitespace tokens), compile a
         Python regex equivalent of the Flex pattern and find unique matches in the
         input text.
     4. Write the matches to `{token_name}_tokens.txt` in the analyzer directory.
    """

    def_path = Path(def_file)
    output_dir = def_path.parent

    input_file = find_input_file(def_path)
    if input_file is None or not input_file.exists():
        print(f'No input file found for {def_path.name}; skipping token generation.')
        return

    try:
        text = input_file.read_text()
    except OSError as exc:
        print(f'Could not read input file {input_file}: {exc}')
        return

    # Remove previously generated token files in this analyzer directory
    removed = 0
    for token_file in output_dir.glob('*_tokens.txt'):
        try:
            token_file.unlink()
            removed += 1
        except OSError as exc:
            print(f'Warning: could not remove {token_file.name}: {exc}')

    if removed:
        print(f'Removed {removed} old token file(s) from {output_dir}')

    skip_tokens = {
        'WHITESPACE', 'NEWLINE', 'SPACE', 'TAB',
        'COMMA', 'DOT', 'COLON', 'SEMICOLON',
        'LPAREN', 'RPAREN', 'LBRACE', 'RBRACE',
        'LBRACKET', 'RBRACKET', 'QUOTE', 'DQUOTE',
        'PLUS', 'MINUS', 'TIMES', 'DIVIDE', 'STAR',
        'EQUALS', 'DASH', 'SLASH', 'BACKSLASH',
        'PERCENT', 'DOLLAR', 'AT', 'HASH', 'EXCLAIM',
        'QUESTION', 'AMPERSAND', 'PIPE', 'CARET',
        'TILDE', 'BACKTICK', 'UNDERSCORE',
        'LETTER', 'DIGIT', 'WORD', 'CHAR'
    }

    max_samples = 50
    files_created = 0

    for rule in lex_rules:
        token_name = rule.token_name

        if token_name.upper() in skip_tokens:
            continue

        pattern = flex_regex_to_python(rule.regex)
        try:
            regex = re.compile(pattern, re.MULTILINE)
        except re.error as exc:
            print(f'Warning: could not compile regex for token {token_name}: {exc}')
            continue

        matches = []
        seen = set()
        for match in regex.finditer(text):
            value = match.group(0)
            if not value:
                continue
            if value in seen:
                continue
            matches.append(value)
            seen.add(value)
            if len(matches) >= max_samples:
                break

        if not matches:
            continue

        filename = f"{token_name.lower()}_tokens.txt"
        filepath = output_dir / filename
        try:
            with open(filepath, 'w') as f:
                for value in matches:
                    f.write(f"{value}\n")
        except OSError as exc:
            print(f'Warning: could not write {filepath.name}: {exc}')
            continue

        files_created += 1

    if files_created > 0:
        print(f'Generated {files_created} token example file(s) in {output_dir} from {input_file.name}')
    else:
        print(f'No token data generated from {input_file.name}')

def main():
    if len(sys.argv) != 2:
        print('Usage: generator.py <input.def>')
        sys.exit(1)
    
    def_file = sys.argv[1]
    
    print(f'Parsing {def_file}...')
    lex_rules, grammar_rules = parse_def_file(def_file)
    
    print(f'Found {len(lex_rules)} lexer rules and {len(grammar_rules)} grammar rules')
    
    print('Generating lexer.l...')
    generate_lexer(lex_rules, 'lexer.l')
    
    print('Generating parser.y...')
    generate_parser(lex_rules, grammar_rules, 'parser.y')
    
    print('Generating token example files...')
    generate_token_files(lex_rules, def_file)
    
    print('Generation complete!')

if __name__ == '__main__':
    main()

File: generator.c
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef enum {
    SECTION_NONE = 0,
    SECTION_LEX,
    SECTION_YACC
} Section;

typedef enum {
    LEX_ITEM_RULE = 0,
    LEX_ITEM_COMMENT,
    LEX_ITEM_BLANK
} LexItemType;

typedef struct {
    LexItemType type;
    char *name;
    char *pattern;
    char *text;
} LexItem;

typedef enum {
    GRAMMAR_ITEM_RULE = 0,
    GRAMMAR_ITEM_COMMENT,
    GRAMMAR_ITEM_BLANK
} GrammarItemType;

typedef struct {
    char **alternatives;
    size_t alt_count;
    size_t alt_capacity;
    char *lhs;
} GrammarRule;

typedef struct {
    GrammarItemType type;
    GrammarRule rule;
    char *text;
} GrammarItem;

typedef struct {
    char **items;
    size_t count;
    size_t capacity;
} TokenList;

static void handle_allocation_failure(void)
{
    fprintf(stderr, "Out of memory\n");
    exit(EXIT_FAILURE);
}

static char *dup_string(const char *input)
{
    size_t length = input ? strlen(input) : 0;
    char *copy = (char *)malloc(length + 1);
    if (!copy) {
        handle_allocation_failure();
    }
    if (length > 0 && input) {
        memcpy(copy, input, length);
    }
    copy[length] = '\0';
    return copy;
}

static char *trim(char *text)
{
    if (!text) {
        return NULL;
    }

    while (*text && isspace((unsigned char)*text)) {
        text++;
    }

    if (*text == '\0') {
        return text;
    }

    char *end = text + strlen(text) - 1;
    while (end > text && isspace((unsigned char)*end)) {
        *end = '\0';
        end--;
    }
    return text;
}

static void ensure_lex_capacity(LexItem **items, size_t *capacity, size_t needed)
{
    if (*capacity >= needed) {
        return;
    }
    size_t new_capacity = *capacity ? *capacity * 2 : 16;
    while (new_capacity < needed) {
        new_capacity *= 2;
    }
    LexItem *resized = (LexItem *)realloc(*items, new_capacity * sizeof(LexItem));
    if (!resized) {
        handle_allocation_failure();
    }
    *items = resized;
    *capacity = new_capacity;
}

static void ensure_grammar_capacity(GrammarItem **items, size_t *capacity, size_t needed)
{
    if (*capacity >= needed) {
        return;
    }
    size_t new_capacity = *capacity ? *capacity * 2 : 16;
    while (new_capacity < needed) {
        new_capacity *= 2;
    }
    GrammarItem *resized = (GrammarItem *)realloc(*items, new_capacity * sizeof(GrammarItem));
    if (!resized) {
        handle_allocation_failure();
    }
    *items = resized;
    *capacity = new_capacity;
}

static void ensure_alt_capacity(GrammarRule *rule, size_t needed)
{
    if (rule->alt_capacity >= needed) {
        return;
    }
    size_t new_capacity = rule->alt_capacity ? rule->alt_capacity * 2 : 4;
    while (new_capacity < needed) {
        new_capacity *= 2;
    }
    char **resized = (char **)realloc(rule->alternatives, new_capacity * sizeof(char *));
    if (!resized) {
        handle_allocation_failure();
    }
    rule->alternatives = resized;
    rule->alt_capacity = new_capacity;
}

static void ensure_token_capacity(TokenList *tokens, size_t needed)
{
    if (tokens->capacity >= needed) {
        return;
    }
    size_t new_capacity = tokens->capacity ? tokens->capacity * 2 : 16;
    while (new_capacity < needed) {
        new_capacity *= 2;
    }
    char **resized = (char **)realloc(tokens->items, new_capacity * sizeof(char *));
    if (!resized) {
        handle_allocation_failure();
    }
    tokens->items = resized;
    tokens->capacity = new_capacity;
}

static int equals_ignore_case(const char *a, const char *b)
{
    if (!a || !b) {
        return 0;
    }
    while (*a && *b) {
        if (tolower((unsigned char)*a) != tolower((unsigned char)*b)) {
            return 0;
        }
        a++;
        b++;
    }
    return *a == '\0' && *b == '\0';
}

static int contains_ignore_case(const char *haystack, const char *needle)
{
    if (!haystack || !needle) {
        return 0;
    }
    size_t needle_len = strlen(needle);
    if (needle_len == 0) {
        return 1;
    }
    for (const char *cursor = haystack; *cursor; ++cursor) {
        if (tolower((unsigned char)*cursor) == tolower((unsigned char)*needle)) {
            size_t index = 0;
            while (needle[index] && cursor[index] && tolower((unsigned char)cursor[index]) == tolower((unsigned char)needle[index])) {
                index++;
            }
            if (needle[index] == '\0') {
                return 1;
            }
        }
    }
    return 0;
}

static int is_skip_token(const char *name)
{
    return equals_ignore_case(name, "WHITESPACE") ||
           equals_ignore_case(name, "WS") ||
            contains_ignore_case(name, "WHITESPACE") ||
           contains_ignore_case(name, "COMMENT");
}

static int is_number_token(const char *name, const char *pattern)
{
    if (equals_ignore_case(name, "NUMBER") || equals_ignore_case(name, "INT") || equals_ignore_case(name, "INTEGER")) {
        return 1;
    }
    if (pattern && (strstr(pattern, "0-9") || strstr(pattern, "[:digit:]"))) {
        return 1;
    }
    return 0;
}

static int token_exists(const TokenList *tokens, const char *name)
{
    for (size_t i = 0; i < tokens->count; ++i) {
        if (equals_ignore_case(tokens->items[i], name)) {
            return 1;
        }
    }
    return 0;
}

static void add_token(TokenList *tokens, const char *name)
{
    if (!tokens || !name || is_skip_token(name)) {
        return;
    }
    if (token_exists(tokens, name)) {
        return;
    }
    ensure_token_capacity(tokens, tokens->count + 1);
    tokens->items[tokens->count++] = dup_string(name);
}

static void grammar_rule_add_alt(GrammarRule *rule, const char *rhs)
{
    ensure_alt_capacity(rule, rule->alt_count + 1);
    rule->alternatives[rule->alt_count++] = dup_string(rhs);
}

static void free_lex_items(LexItem *items, size_t count)
{
    for (size_t i = 0; i < count; ++i) {
        free(items[i].name);
        free(items[i].pattern);
        free(items[i].text);
    }
    free(items);
}

static void free_grammar_items(GrammarItem *items, size_t count)
{
    for (size_t i = 0; i < count; ++i) {
        if (items[i].type == GRAMMAR_ITEM_RULE) {
            for (size_t j = 0; j < items[i].rule.alt_count; ++j) {
                free(items[i].rule.alternatives[j]);
            }
            free(items[i].rule.alternatives);
            free(items[i].rule.lhs);
        }
        free(items[i].text);
    }
    free(items);
}

static void free_tokens(TokenList *tokens)
{
    for (size_t i = 0; i < tokens->count; ++i) {
        free(tokens->items[i]);
    }
    free(tokens->items);
}

static void write_lexer(const char *path, const LexItem *items, size_t count)
{
    FILE *out = fopen(path, "w");
    if (!out) {
        perror("Failed to open lexer.l for writing");
        exit(EXIT_FAILURE);
    }

    fprintf(out, "%%{\n");
    fprintf(out, "#include \"y.tab.h\"\n");
    fprintf(out, "#include <stdio.h>\n");
    fprintf(out, "#include <stdlib.h>\n");
    fprintf(out, "#include <string.h>\n");
    fprintf(out, "%%}\n\n");

    fprintf(out, "%%option noyywrap\n\n");
    fputs("%%\n\n", out);

    for (size_t i = 0; i < count; ++i) {
        const LexItem *item = &items[i];
        switch (item->type) {
        case LEX_ITEM_BLANK:
            fputc('\n', out);
            break;
        case LEX_ITEM_COMMENT:
            fprintf(out, "%s\n", item->text ? item->text : "");
            break;
        case LEX_ITEM_RULE:
            if (!item->name || !item->pattern) {
                break;
            }
            if (is_skip_token(item->name)) {
                fprintf(out, "%s\t{ /* skip %s */ }\n", item->pattern, item->name);
            } else if (is_number_token(item->name, item->pattern)) {
                fprintf(out, "%s\t{ yylval = (int)strtol(yytext, NULL, 10); return %s; }\n", item->pattern, item->name);
            } else {
                fprintf(out, "%s\t{ return %s; }\n", item->pattern, item->name);
            }
            break;
        }
    }

    fputs(".\t{ return yytext[0]; }\n", out);
    fputs("%%\n", out);

    fclose(out);
}

static void write_parser(const char *path, const GrammarItem *items, size_t count, const TokenList *tokens)
{
    FILE *out = fopen(path, "w");
    if (!out) {
        perror("Failed to open parser.y for writing");
        exit(EXIT_FAILURE);
    }

    fprintf(out, "%%{\n");
    fprintf(out, "#include <stdio.h>\n");
    fprintf(out, "#include <stdlib.h>\n");
    fprintf(out, "extern int yylex(void);\n");
    fprintf(out, "extern FILE *yyin;\n");
    fprintf(out, "void yyerror(const char *message);\n");
    fprintf(out, "%%}\n\n");

    if (tokens && tokens->count > 0) {
        fputs("%token", out);
        for (size_t i = 0; i < tokens->count; ++i) {
            fprintf(out, " %s", tokens->items[i]);
        }
        fputs("\n\n", out);
    }

    fputs("%%\n\n", out);

    for (size_t i = 0; i < count; ++i) {
        const GrammarItem *item = &items[i];
        switch (item->type) {
        case GRAMMAR_ITEM_BLANK:
            fputc('\n', out);
            break;
        case GRAMMAR_ITEM_COMMENT:
            fprintf(out, "%s\n", item->text ? item->text : "");
            break;
        case GRAMMAR_ITEM_RULE:
            if (!item->rule.lhs || item->rule.alt_count == 0) {
                break;
            }
            fprintf(out, "%s:\n", item->rule.lhs);
            for (size_t alt = 0; alt < item->rule.alt_count; ++alt) {
                if (alt == 0) {
                    fprintf(out, "    %s\n", item->rule.alternatives[alt]);
                } else {
                    fprintf(out, "  | %s\n", item->rule.alternatives[alt]);
                }
            }
            fputs(";\n\n", out);
            break;
        }
    }

    fputs("%%\n\n", out);

    fputs("int main(int argc, char **argv)\n", out);
    fputs("{\n", out);
    fputs("    if (argc != 2) {\n", out);
    fputs("        fprintf(stderr, \"Usage: %s <source-file>\\n\", argc > 0 ? argv[0] : \"parser\");\n", out);
    fputs("        return EXIT_FAILURE;\n", out);
    fputs("    }\n\n", out);
    fputs("    FILE *input = fopen(argv[1], \"r\");\n", out);
    fputs("    if (!input) {\n", out);
    fputs("        perror(\"Failed to open source file\");\n", out);
    fputs("        return EXIT_FAILURE;\n", out);
    fputs("    }\n\n", out);
    fputs("    yyin = input;\n", out);
    fputs("    int status = yyparse();\n", out);
    fputs("    fclose(input);\n\n", out);
    fputs("    return status == 0 ? EXIT_SUCCESS : EXIT_FAILURE;\n", out);
    fputs("}\n\n", out);

    fputs("void yyerror(const char *message)\n", out);
    fputs("{\n", out);
    fputs("    fprintf(stderr, \"Parse error: %s\\n\", message);\n", out);
    fputs("}\n", out);

    fclose(out);
}

int main(int argc, char **argv)
{
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <language.def>\n", argc > 0 ? argv[0] : "generator");
        return EXIT_FAILURE;
    }

    const char *definition_path = argv[1];
    FILE *definition = fopen(definition_path, "r");
    if (!definition) {
        perror("Failed to open definition file");
        return EXIT_FAILURE;
    }

    LexItem *lex_items = NULL;
    size_t lex_count = 0;
    size_t lex_capacity = 0;

    GrammarItem *grammar_items = NULL;
    size_t grammar_count = 0;
    size_t grammar_capacity = 0;

    TokenList tokens = {0};

    Section section = SECTION_NONE;
    size_t last_rule_index = (size_t)(-1);
    char line_buffer[4096];

    while (fgets(line_buffer, sizeof(line_buffer), definition)) {
        size_t length = strlen(line_buffer);
        while (length > 0 && (line_buffer[length - 1] == '\n' || line_buffer[length - 1] == '\r')) {
            line_buffer[--length] = '\0';
        }

        char *line_copy = dup_string(line_buffer);
        char *trimmed = trim(line_copy);

        if (strcmp(trimmed, "%%LEX") == 0) {
            section = SECTION_LEX;
            free(line_copy);
            continue;
        }

        if (strcmp(trimmed, "%%YACC") == 0) {
            section = SECTION_YACC;
            free(line_copy);
            continue;
        }

        if (section == SECTION_NONE) {
            free(line_copy);
            continue;
        }

        if (*trimmed == '\0') {
            if (section == SECTION_LEX) {
                ensure_lex_capacity(&lex_items, &lex_capacity, lex_count + 1);
                lex_items[lex_count].type = LEX_ITEM_BLANK;
                lex_items[lex_count].name = NULL;
                lex_items[lex_count].pattern = NULL;
                lex_items[lex_count].text = NULL;
                lex_count++;
            } else if (section == SECTION_YACC) {
                ensure_grammar_capacity(&grammar_items, &grammar_capacity, grammar_count + 1);
                grammar_items[grammar_count].type = GRAMMAR_ITEM_BLANK;
                grammar_items[grammar_count].text = NULL;
                grammar_count++;
            }
            free(line_copy);
            continue;
        }

        if (trimmed[0] == '/' && trimmed[1] == '/') {
            if (section == SECTION_LEX) {
                ensure_lex_capacity(&lex_items, &lex_capacity, lex_count + 1);
                lex_items[lex_count].type = LEX_ITEM_COMMENT;
                lex_items[lex_count].name = NULL;
                lex_items[lex_count].pattern = NULL;
                lex_items[lex_count].text = dup_string(trimmed);
                lex_count++;
            } else if (section == SECTION_YACC) {
                ensure_grammar_capacity(&grammar_items, &grammar_capacity, grammar_count + 1);
                grammar_items[grammar_count].type = GRAMMAR_ITEM_COMMENT;
                grammar_items[grammar_count].text = dup_string(trimmed);
                grammar_count++;
            }
            free(line_copy);
            continue;
        }

        if (section == SECTION_LEX) {
            char *rule_copy = dup_string(trimmed);
            char *cursor = rule_copy;
            while (*cursor && !isspace((unsigned char)*cursor)) {
                cursor++;
            }
            if (*cursor == '\0') {
                fprintf(stderr, "Invalid lexer rule: %s\n", trimmed);
                free(rule_copy);
                free(line_copy);
                free_lex_items(lex_items, lex_count);
                free_grammar_items(grammar_items, grammar_count);
                free_tokens(&tokens);
                fclose(definition);
                return EXIT_FAILURE;
            }
            *cursor = '\0';
            char *pattern = trim(cursor + 1);

            ensure_lex_capacity(&lex_items, &lex_capacity, lex_count + 1);
            lex_items[lex_count].type = LEX_ITEM_RULE;
            lex_items[lex_count].name = dup_string(rule_copy);
            lex_items[lex_count].pattern = dup_string(pattern);
            lex_items[lex_count].text = NULL;
            add_token(&tokens, rule_copy);
            lex_count++;

            free(rule_copy);
            free(line_copy);
            continue;
        }

        if (section == SECTION_YACC) {
            char *arrow = strstr(trimmed, "->");
            if (arrow) {
                *arrow = '\0';
                char *lhs_text = trim(trimmed);
                char *rhs_text = trim(arrow + 2);

                ensure_grammar_capacity(&grammar_items, &grammar_capacity, grammar_count + 1);
                GrammarItem *item = &grammar_items[grammar_count];
                item->type = GRAMMAR_ITEM_RULE;
                item->text = NULL;
                item->rule.lhs = dup_string(lhs_text);
                item->rule.alternatives = NULL;
                item->rule.alt_count = 0;
                item->rule.alt_capacity = 0;
                grammar_rule_add_alt(&item->rule, rhs_text);
                last_rule_index = grammar_count;
                grammar_count++;
                free(line_copy);
                continue;
            }

            if (trimmed[0] == '|') {
                if (last_rule_index == (size_t)(-1)) {
                    fprintf(stderr, "Dangling alternative without a preceding rule\n");
                    free(line_copy);
                    free_lex_items(lex_items, lex_count);
                    free_grammar_items(grammar_items, grammar_count);
                    free_tokens(&tokens);
                    fclose(definition);
                    return EXIT_FAILURE;
                }
                char *rhs_text = trim(trimmed + 1);
                grammar_rule_add_alt(&grammar_items[last_rule_index].rule, rhs_text);
                free(line_copy);
                continue;
            }

            /* Any other line is treated as a comment in the grammar section. */
            ensure_grammar_capacity(&grammar_items, &grammar_capacity, grammar_count + 1);
            grammar_items[grammar_count].type = GRAMMAR_ITEM_COMMENT;
            grammar_items[grammar_count].text = dup_string(trimmed);
            grammar_count++;
            free(line_copy);
            continue;
        }

        free(line_copy);
    }

    fclose(definition);

    if (lex_count == 0 || grammar_count == 0) {
        fprintf(stderr, "Definition file must contain non-empty %%LEX and %%YACC sections\n");
        free_lex_items(lex_items, lex_count);
        free_grammar_items(grammar_items, grammar_count);
        free_tokens(&tokens);
        return EXIT_FAILURE;
    }

    write_lexer("lexer.l", lex_items, lex_count);
    write_parser("parser.y", grammar_items, grammar_count, &tokens);

    free_lex_items(lex_items, lex_count);
    free_grammar_items(grammar_items, grammar_count);
    free_tokens(&tokens);

    return EXIT_SUCCESS;
}
